### PR TITLE
ODIN II: re-modelled for loop unrolling, module instantiation

### PR DESCRIPTION
**Pull request by ademmings**
#### Description
- For loops are now being unrolled during AST elaboration as FOR nodes are found, rather than in its own sweep
- Module instantiations are moved to a later phase after generate constructs are resolved, so only the instances that are used are instantiated
- A bug that allowed module instances of the same name to occur was fixed

#### How Has This Been Tested?
- Odin pre-commit regression suite

#### Types of changes

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by jeanlego** [comment]
could you fix the merge conflict and rerun the regression suite? @j-b-1-7 included new Valgrind ones

**Comment by ademmings** [comment]
@jeanlego fixed. I actually had to remove the free in newFunctionInstance() because what used to be discarded and leaked is now being used, since I removed the make_module_param_name() function

File: ODIN_II/SRC/enum_str.cpp [hunk 1]
```diff
@@ -167,6 +167,7 @@ const char *ids_STR []=
 	"IF",
 	"IF_Q",
 	"FOR",
+	"UNROLLED_FOR",
```
Comment (jeanlego) [inline ODIN_II/SRC/enum_str.cpp]: Is this used ? otherwise I would remove

File: ODIN_II/SRC/include/odin_types.h [hunk 2]
```diff
@@ -303,6 +303,7 @@ enum ids
 	IF,
 	IF_Q,
 	FOR,
+	UNROLLED_FOR,
```
Comment (jeanlego) [inline ODIN_II/SRC/include/odin_types.h]: is this used?

File: ODIN_II/regression_test/benchmark/verilog/syntax/both_ram.v [hunk 3]
```diff
@@ -60,7 +60,7 @@ dual_port_ram inst2(
   .addr2(address2)
 );
 
-single_port_ram inst1(
+single_port_ram inst3(
```
Comment (jeanlego) [inline ODIN_II/regression_test/benchmark/verilog/syntax/both_ram.v]: is this change needed ?

File: ODIN_II/SRC/ast_loop_unroll.cpp [hunk 4]
```diff
@@ -599,109 +393,31 @@ ast_node_t* dup_and_fill_body(ast_node_t *ast_module, ast_node_t* body, ast_node
 			} 
 			else if(child->type == MODULE_INSTANCE && child->children[0]->type != MODULE_INSTANCE)
```
Comment (jeanlego) [inline ODIN_II/SRC/ast_loop_unroll.cpp]: set a boolean (i.e. unrolled) when this condition is met and then use the boolean for the next time you need to check that it was not unrolled.

File: ODIN_II/SRC/ast_loop_unroll.cpp [hunk 5]
```diff
@@ -599,109 +393,31 @@ ast_node_t* dup_and_fill_body(ast_node_t *ast_module, ast_node_t* body, ast_node
...
+
+				/* find and replace iteration symbol for port connections and parameters */
+				ast_node_t *named_instance = child->children[1];
+				copy->children[i]->children[1] = dup_and_fill_body(ast_module, named_instance, pre, value, error_code);
+				free_whole_tree(named_instance);
 			} 
 
 			if(child && child->num_children > 0)
 			{
 				if (child->type == MODULE_INSTANCE && child->children[0]->type != MODULE_INSTANCE)
```
Comment (jeanlego) [inline ODIN_II/SRC/ast_loop_unroll.cpp]: use the said boolean here, and use if(!unrolled) then ... rather then if(unrolled) then do nothing else ...

File: ODIN_II/regression_test/benchmark/verilog/syntax/both_ram.v (hunk 3 above)
Comment (ademmings) [reply ODIN_II/regression_test/benchmark/verilog/syntax/both_ram.v]: yes
